// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: v1/contact.proto

package api_v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ContactListRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactListRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactListRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactListRequestMultiError, or nil if none found.
func (m *ContactListRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactListRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactListRequestMultiError(errors)
	}

	return nil
}

// ContactListRequestMultiError is an error wrapping multiple validation errors
// returned by ContactListRequest.ValidateAll() if the designated constraints
// aren't met.
type ContactListRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactListRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactListRequestMultiError) AllErrors() []error { return m }

// ContactListRequestValidationError is the validation error returned by
// ContactListRequest.Validate if the designated constraints aren't met.
type ContactListRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactListRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactListRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactListRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactListRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactListRequestValidationError) ErrorName() string {
	return "ContactListRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactListRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactListRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactListRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactListRequestValidationError{}

// Validate checks the field values on ContactListResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactListResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactListResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactListResponseMultiError, or nil if none found.
func (m *ContactListResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactListResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetItems() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ContactListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ContactListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactListResponseValidationError{
					field:  fmt.Sprintf("Items[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ContactListResponseMultiError(errors)
	}

	return nil
}

// ContactListResponseMultiError is an error wrapping multiple validation
// errors returned by ContactListResponse.ValidateAll() if the designated
// constraints aren't met.
type ContactListResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactListResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactListResponseMultiError) AllErrors() []error { return m }

// ContactListResponseValidationError is the validation error returned by
// ContactListResponse.Validate if the designated constraints aren't met.
type ContactListResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactListResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactListResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactListResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactListResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactListResponseValidationError) ErrorName() string {
	return "ContactListResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactListResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactListResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactListResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactListResponseValidationError{}

// Validate checks the field values on ContactDeleteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactDeleteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactDeleteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactDeleteRequestMultiError, or nil if none found.
func (m *ContactDeleteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactDeleteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for FriendId

	if len(errors) > 0 {
		return ContactDeleteRequestMultiError(errors)
	}

	return nil
}

// ContactDeleteRequestMultiError is an error wrapping multiple validation
// errors returned by ContactDeleteRequest.ValidateAll() if the designated
// constraints aren't met.
type ContactDeleteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactDeleteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactDeleteRequestMultiError) AllErrors() []error { return m }

// ContactDeleteRequestValidationError is the validation error returned by
// ContactDeleteRequest.Validate if the designated constraints aren't met.
type ContactDeleteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactDeleteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactDeleteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactDeleteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactDeleteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactDeleteRequestValidationError) ErrorName() string {
	return "ContactDeleteRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactDeleteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactDeleteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactDeleteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactDeleteRequestValidationError{}

// Validate checks the field values on ContactDeleteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactDeleteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactDeleteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactDeleteResponseMultiError, or nil if none found.
func (m *ContactDeleteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactDeleteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactDeleteResponseMultiError(errors)
	}

	return nil
}

// ContactDeleteResponseMultiError is an error wrapping multiple validation
// errors returned by ContactDeleteResponse.ValidateAll() if the designated
// constraints aren't met.
type ContactDeleteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactDeleteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactDeleteResponseMultiError) AllErrors() []error { return m }

// ContactDeleteResponseValidationError is the validation error returned by
// ContactDeleteResponse.Validate if the designated constraints aren't met.
type ContactDeleteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactDeleteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactDeleteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactDeleteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactDeleteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactDeleteResponseValidationError) ErrorName() string {
	return "ContactDeleteResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactDeleteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactDeleteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactDeleteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactDeleteResponseValidationError{}

// Validate checks the field values on ContactDetailRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactDetailRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactDetailRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactDetailRequestMultiError, or nil if none found.
func (m *ContactDetailRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactDetailRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	if len(errors) > 0 {
		return ContactDetailRequestMultiError(errors)
	}

	return nil
}

// ContactDetailRequestMultiError is an error wrapping multiple validation
// errors returned by ContactDetailRequest.ValidateAll() if the designated
// constraints aren't met.
type ContactDetailRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactDetailRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactDetailRequestMultiError) AllErrors() []error { return m }

// ContactDetailRequestValidationError is the validation error returned by
// ContactDetailRequest.Validate if the designated constraints aren't met.
type ContactDetailRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactDetailRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactDetailRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactDetailRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactDetailRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactDetailRequestValidationError) ErrorName() string {
	return "ContactDetailRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactDetailRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactDetailRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactDetailRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactDetailRequestValidationError{}

// Validate checks the field values on ContactDetailResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactDetailResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactDetailResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactDetailResponseMultiError, or nil if none found.
func (m *ContactDetailResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactDetailResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Username

	// no validation rules for Avatar

	// no validation rules for Name

	// no validation rules for Surname

	// no validation rules for Gender

	// no validation rules for About

	// no validation rules for FriendStatus

	// no validation rules for GroupId

	// no validation rules for IsBot

	if len(errors) > 0 {
		return ContactDetailResponseMultiError(errors)
	}

	return nil
}

// ContactDetailResponseMultiError is an error wrapping multiple validation
// errors returned by ContactDetailResponse.ValidateAll() if the designated
// constraints aren't met.
type ContactDetailResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactDetailResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactDetailResponseMultiError) AllErrors() []error { return m }

// ContactDetailResponseValidationError is the validation error returned by
// ContactDetailResponse.Validate if the designated constraints aren't met.
type ContactDetailResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactDetailResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactDetailResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactDetailResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactDetailResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactDetailResponseValidationError) ErrorName() string {
	return "ContactDetailResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactDetailResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactDetailResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactDetailResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactDetailResponseValidationError{}

// Validate checks the field values on ContactChangeGroupRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactChangeGroupRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactChangeGroupRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactChangeGroupRequestMultiError, or nil if none found.
func (m *ContactChangeGroupRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactChangeGroupRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	// no validation rules for GroupId

	if len(errors) > 0 {
		return ContactChangeGroupRequestMultiError(errors)
	}

	return nil
}

// ContactChangeGroupRequestMultiError is an error wrapping multiple validation
// errors returned by ContactChangeGroupRequest.ValidateAll() if the
// designated constraints aren't met.
type ContactChangeGroupRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactChangeGroupRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactChangeGroupRequestMultiError) AllErrors() []error { return m }

// ContactChangeGroupRequestValidationError is the validation error returned by
// ContactChangeGroupRequest.Validate if the designated constraints aren't met.
type ContactChangeGroupRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactChangeGroupRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactChangeGroupRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactChangeGroupRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactChangeGroupRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactChangeGroupRequestValidationError) ErrorName() string {
	return "ContactChangeGroupRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactChangeGroupRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactChangeGroupRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactChangeGroupRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactChangeGroupRequestValidationError{}

// Validate checks the field values on ContactChangeGroupResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactChangeGroupResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactChangeGroupResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactChangeGroupResponseMultiError, or nil if none found.
func (m *ContactChangeGroupResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactChangeGroupResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactChangeGroupResponseMultiError(errors)
	}

	return nil
}

// ContactChangeGroupResponseMultiError is an error wrapping multiple
// validation errors returned by ContactChangeGroupResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactChangeGroupResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactChangeGroupResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactChangeGroupResponseMultiError) AllErrors() []error { return m }

// ContactChangeGroupResponseValidationError is the validation error returned
// by ContactChangeGroupResponse.Validate if the designated constraints aren't met.
type ContactChangeGroupResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactChangeGroupResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactChangeGroupResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactChangeGroupResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactChangeGroupResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactChangeGroupResponseValidationError) ErrorName() string {
	return "ContactChangeGroupResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactChangeGroupResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactChangeGroupResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactChangeGroupResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactChangeGroupResponseValidationError{}

// Validate checks the field values on ContactRequestCreateRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestCreateRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestCreateRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactRequestCreateRequestMultiError, or nil if none found.
func (m *ContactRequestCreateRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestCreateRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for FriendId

	if len(errors) > 0 {
		return ContactRequestCreateRequestMultiError(errors)
	}

	return nil
}

// ContactRequestCreateRequestMultiError is an error wrapping multiple
// validation errors returned by ContactRequestCreateRequest.ValidateAll() if
// the designated constraints aren't met.
type ContactRequestCreateRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestCreateRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestCreateRequestMultiError) AllErrors() []error { return m }

// ContactRequestCreateRequestValidationError is the validation error returned
// by ContactRequestCreateRequest.Validate if the designated constraints
// aren't met.
type ContactRequestCreateRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestCreateRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestCreateRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestCreateRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestCreateRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestCreateRequestValidationError) ErrorName() string {
	return "ContactRequestCreateRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestCreateRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestCreateRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestCreateRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestCreateRequestValidationError{}

// Validate checks the field values on ContactRequestCreateResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestCreateResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestCreateResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactRequestCreateResponseMultiError, or nil if none found.
func (m *ContactRequestCreateResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestCreateResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactRequestCreateResponseMultiError(errors)
	}

	return nil
}

// ContactRequestCreateResponseMultiError is an error wrapping multiple
// validation errors returned by ContactRequestCreateResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactRequestCreateResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestCreateResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestCreateResponseMultiError) AllErrors() []error { return m }

// ContactRequestCreateResponseValidationError is the validation error returned
// by ContactRequestCreateResponse.Validate if the designated constraints
// aren't met.
type ContactRequestCreateResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestCreateResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestCreateResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestCreateResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestCreateResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestCreateResponseValidationError) ErrorName() string {
	return "ContactRequestCreateResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestCreateResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestCreateResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestCreateResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestCreateResponseValidationError{}

// Validate checks the field values on ContactRequestAcceptRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestAcceptRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestAcceptRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactRequestAcceptRequestMultiError, or nil if none found.
func (m *ContactRequestAcceptRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestAcceptRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ApplyId

	if len(errors) > 0 {
		return ContactRequestAcceptRequestMultiError(errors)
	}

	return nil
}

// ContactRequestAcceptRequestMultiError is an error wrapping multiple
// validation errors returned by ContactRequestAcceptRequest.ValidateAll() if
// the designated constraints aren't met.
type ContactRequestAcceptRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestAcceptRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestAcceptRequestMultiError) AllErrors() []error { return m }

// ContactRequestAcceptRequestValidationError is the validation error returned
// by ContactRequestAcceptRequest.Validate if the designated constraints
// aren't met.
type ContactRequestAcceptRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestAcceptRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestAcceptRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestAcceptRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestAcceptRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestAcceptRequestValidationError) ErrorName() string {
	return "ContactRequestAcceptRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestAcceptRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestAcceptRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestAcceptRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestAcceptRequestValidationError{}

// Validate checks the field values on ContactRequestAcceptResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestAcceptResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestAcceptResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactRequestAcceptResponseMultiError, or nil if none found.
func (m *ContactRequestAcceptResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestAcceptResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactRequestAcceptResponseMultiError(errors)
	}

	return nil
}

// ContactRequestAcceptResponseMultiError is an error wrapping multiple
// validation errors returned by ContactRequestAcceptResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactRequestAcceptResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestAcceptResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestAcceptResponseMultiError) AllErrors() []error { return m }

// ContactRequestAcceptResponseValidationError is the validation error returned
// by ContactRequestAcceptResponse.Validate if the designated constraints
// aren't met.
type ContactRequestAcceptResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestAcceptResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestAcceptResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestAcceptResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestAcceptResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestAcceptResponseValidationError) ErrorName() string {
	return "ContactRequestAcceptResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestAcceptResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestAcceptResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestAcceptResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestAcceptResponseValidationError{}

// Validate checks the field values on ContactRequestDeclineRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestDeclineRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestDeclineRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactRequestDeclineRequestMultiError, or nil if none found.
func (m *ContactRequestDeclineRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestDeclineRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ApplyId

	if len(errors) > 0 {
		return ContactRequestDeclineRequestMultiError(errors)
	}

	return nil
}

// ContactRequestDeclineRequestMultiError is an error wrapping multiple
// validation errors returned by ContactRequestDeclineRequest.ValidateAll() if
// the designated constraints aren't met.
type ContactRequestDeclineRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestDeclineRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestDeclineRequestMultiError) AllErrors() []error { return m }

// ContactRequestDeclineRequestValidationError is the validation error returned
// by ContactRequestDeclineRequest.Validate if the designated constraints
// aren't met.
type ContactRequestDeclineRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestDeclineRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestDeclineRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestDeclineRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestDeclineRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestDeclineRequestValidationError) ErrorName() string {
	return "ContactRequestDeclineRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestDeclineRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestDeclineRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestDeclineRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestDeclineRequestValidationError{}

// Validate checks the field values on ContactRequestDeclineResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestDeclineResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestDeclineResponse with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// ContactRequestDeclineResponseMultiError, or nil if none found.
func (m *ContactRequestDeclineResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestDeclineResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactRequestDeclineResponseMultiError(errors)
	}

	return nil
}

// ContactRequestDeclineResponseMultiError is an error wrapping multiple
// validation errors returned by ContactRequestDeclineResponse.ValidateAll()
// if the designated constraints aren't met.
type ContactRequestDeclineResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestDeclineResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestDeclineResponseMultiError) AllErrors() []error { return m }

// ContactRequestDeclineResponseValidationError is the validation error
// returned by ContactRequestDeclineResponse.Validate if the designated
// constraints aren't met.
type ContactRequestDeclineResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestDeclineResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestDeclineResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestDeclineResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestDeclineResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestDeclineResponseValidationError) ErrorName() string {
	return "ContactRequestDeclineResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestDeclineResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestDeclineResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestDeclineResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestDeclineResponseValidationError{}

// Validate checks the field values on ContactRequestListResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestListResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestListResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactRequestListResponseMultiError, or nil if none found.
func (m *ContactRequestListResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestListResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetItems() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ContactRequestListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ContactRequestListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactRequestListResponseValidationError{
					field:  fmt.Sprintf("Items[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ContactRequestListResponseMultiError(errors)
	}

	return nil
}

// ContactRequestListResponseMultiError is an error wrapping multiple
// validation errors returned by ContactRequestListResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactRequestListResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestListResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestListResponseMultiError) AllErrors() []error { return m }

// ContactRequestListResponseValidationError is the validation error returned
// by ContactRequestListResponse.Validate if the designated constraints aren't met.
type ContactRequestListResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestListResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestListResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestListResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestListResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestListResponseValidationError) ErrorName() string {
	return "ContactRequestListResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestListResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestListResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestListResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestListResponseValidationError{}

// Validate checks the field values on ContactGroupCreateRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupCreateRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupCreateRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupCreateRequestMultiError, or nil if none found.
func (m *ContactGroupCreateRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupCreateRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	// no validation rules for Sort

	if len(errors) > 0 {
		return ContactGroupCreateRequestMultiError(errors)
	}

	return nil
}

// ContactGroupCreateRequestMultiError is an error wrapping multiple validation
// errors returned by ContactGroupCreateRequest.ValidateAll() if the
// designated constraints aren't met.
type ContactGroupCreateRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupCreateRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupCreateRequestMultiError) AllErrors() []error { return m }

// ContactGroupCreateRequestValidationError is the validation error returned by
// ContactGroupCreateRequest.Validate if the designated constraints aren't met.
type ContactGroupCreateRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupCreateRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupCreateRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupCreateRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupCreateRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupCreateRequestValidationError) ErrorName() string {
	return "ContactGroupCreateRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupCreateRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupCreateRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupCreateRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupCreateRequestValidationError{}

// Validate checks the field values on ContactGroupCreateResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupCreateResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupCreateResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupCreateResponseMultiError, or nil if none found.
func (m *ContactGroupCreateResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupCreateResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ContactGroupCreateResponseMultiError(errors)
	}

	return nil
}

// ContactGroupCreateResponseMultiError is an error wrapping multiple
// validation errors returned by ContactGroupCreateResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactGroupCreateResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupCreateResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupCreateResponseMultiError) AllErrors() []error { return m }

// ContactGroupCreateResponseValidationError is the validation error returned
// by ContactGroupCreateResponse.Validate if the designated constraints aren't met.
type ContactGroupCreateResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupCreateResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupCreateResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupCreateResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupCreateResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupCreateResponseValidationError) ErrorName() string {
	return "ContactGroupCreateResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupCreateResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupCreateResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupCreateResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupCreateResponseValidationError{}

// Validate checks the field values on ContactGroupUpdateRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupUpdateRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupUpdateRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupUpdateRequestMultiError, or nil if none found.
func (m *ContactGroupUpdateRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupUpdateRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for Sort

	if len(errors) > 0 {
		return ContactGroupUpdateRequestMultiError(errors)
	}

	return nil
}

// ContactGroupUpdateRequestMultiError is an error wrapping multiple validation
// errors returned by ContactGroupUpdateRequest.ValidateAll() if the
// designated constraints aren't met.
type ContactGroupUpdateRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupUpdateRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupUpdateRequestMultiError) AllErrors() []error { return m }

// ContactGroupUpdateRequestValidationError is the validation error returned by
// ContactGroupUpdateRequest.Validate if the designated constraints aren't met.
type ContactGroupUpdateRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupUpdateRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupUpdateRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupUpdateRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupUpdateRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupUpdateRequestValidationError) ErrorName() string {
	return "ContactGroupUpdateRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupUpdateRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupUpdateRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupUpdateRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupUpdateRequestValidationError{}

// Validate checks the field values on ContactGroupUpdateResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupUpdateResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupUpdateResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupUpdateResponseMultiError, or nil if none found.
func (m *ContactGroupUpdateResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupUpdateResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ContactGroupUpdateResponseMultiError(errors)
	}

	return nil
}

// ContactGroupUpdateResponseMultiError is an error wrapping multiple
// validation errors returned by ContactGroupUpdateResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactGroupUpdateResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupUpdateResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupUpdateResponseMultiError) AllErrors() []error { return m }

// ContactGroupUpdateResponseValidationError is the validation error returned
// by ContactGroupUpdateResponse.Validate if the designated constraints aren't met.
type ContactGroupUpdateResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupUpdateResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupUpdateResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupUpdateResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupUpdateResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupUpdateResponseValidationError) ErrorName() string {
	return "ContactGroupUpdateResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupUpdateResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupUpdateResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupUpdateResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupUpdateResponseValidationError{}

// Validate checks the field values on ContactGroupDeleteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupDeleteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupDeleteRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupDeleteRequestMultiError, or nil if none found.
func (m *ContactGroupDeleteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupDeleteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ContactGroupDeleteRequestMultiError(errors)
	}

	return nil
}

// ContactGroupDeleteRequestMultiError is an error wrapping multiple validation
// errors returned by ContactGroupDeleteRequest.ValidateAll() if the
// designated constraints aren't met.
type ContactGroupDeleteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupDeleteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupDeleteRequestMultiError) AllErrors() []error { return m }

// ContactGroupDeleteRequestValidationError is the validation error returned by
// ContactGroupDeleteRequest.Validate if the designated constraints aren't met.
type ContactGroupDeleteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupDeleteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupDeleteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupDeleteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupDeleteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupDeleteRequestValidationError) ErrorName() string {
	return "ContactGroupDeleteRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupDeleteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupDeleteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupDeleteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupDeleteRequestValidationError{}

// Validate checks the field values on ContactGroupDeleteResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupDeleteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupDeleteResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupDeleteResponseMultiError, or nil if none found.
func (m *ContactGroupDeleteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupDeleteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ContactGroupDeleteResponseMultiError(errors)
	}

	return nil
}

// ContactGroupDeleteResponseMultiError is an error wrapping multiple
// validation errors returned by ContactGroupDeleteResponse.ValidateAll() if
// the designated constraints aren't met.
type ContactGroupDeleteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupDeleteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupDeleteResponseMultiError) AllErrors() []error { return m }

// ContactGroupDeleteResponseValidationError is the validation error returned
// by ContactGroupDeleteResponse.Validate if the designated constraints aren't met.
type ContactGroupDeleteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupDeleteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupDeleteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupDeleteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupDeleteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupDeleteResponseValidationError) ErrorName() string {
	return "ContactGroupDeleteResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupDeleteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupDeleteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupDeleteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupDeleteResponseValidationError{}

// Validate checks the field values on ContactGroupSortRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupSortRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupSortRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupSortRequestMultiError, or nil if none found.
func (m *ContactGroupSortRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupSortRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetItems() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ContactGroupSortRequestValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ContactGroupSortRequestValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactGroupSortRequestValidationError{
					field:  fmt.Sprintf("Items[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ContactGroupSortRequestMultiError(errors)
	}

	return nil
}

// ContactGroupSortRequestMultiError is an error wrapping multiple validation
// errors returned by ContactGroupSortRequest.ValidateAll() if the designated
// constraints aren't met.
type ContactGroupSortRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupSortRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupSortRequestMultiError) AllErrors() []error { return m }

// ContactGroupSortRequestValidationError is the validation error returned by
// ContactGroupSortRequest.Validate if the designated constraints aren't met.
type ContactGroupSortRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupSortRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupSortRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupSortRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupSortRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupSortRequestValidationError) ErrorName() string {
	return "ContactGroupSortRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupSortRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupSortRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupSortRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupSortRequestValidationError{}

// Validate checks the field values on ContactGroupSortResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupSortResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupSortResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupSortResponseMultiError, or nil if none found.
func (m *ContactGroupSortResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupSortResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactGroupSortResponseMultiError(errors)
	}

	return nil
}

// ContactGroupSortResponseMultiError is an error wrapping multiple validation
// errors returned by ContactGroupSortResponse.ValidateAll() if the designated
// constraints aren't met.
type ContactGroupSortResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupSortResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupSortResponseMultiError) AllErrors() []error { return m }

// ContactGroupSortResponseValidationError is the validation error returned by
// ContactGroupSortResponse.Validate if the designated constraints aren't met.
type ContactGroupSortResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupSortResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupSortResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupSortResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupSortResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupSortResponseValidationError) ErrorName() string {
	return "ContactGroupSortResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupSortResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupSortResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupSortResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupSortResponseValidationError{}

// Validate checks the field values on ContactGroupListRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupListRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupListRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupListRequestMultiError, or nil if none found.
func (m *ContactGroupListRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupListRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactGroupListRequestMultiError(errors)
	}

	return nil
}

// ContactGroupListRequestMultiError is an error wrapping multiple validation
// errors returned by ContactGroupListRequest.ValidateAll() if the designated
// constraints aren't met.
type ContactGroupListRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupListRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupListRequestMultiError) AllErrors() []error { return m }

// ContactGroupListRequestValidationError is the validation error returned by
// ContactGroupListRequest.Validate if the designated constraints aren't met.
type ContactGroupListRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupListRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupListRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupListRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupListRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupListRequestValidationError) ErrorName() string {
	return "ContactGroupListRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupListRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupListRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupListRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupListRequestValidationError{}

// Validate checks the field values on ContactGroupListResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupListResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupListResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupListResponseMultiError, or nil if none found.
func (m *ContactGroupListResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupListResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetItems() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ContactGroupListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ContactGroupListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactGroupListResponseValidationError{
					field:  fmt.Sprintf("Items[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ContactGroupListResponseMultiError(errors)
	}

	return nil
}

// ContactGroupListResponseMultiError is an error wrapping multiple validation
// errors returned by ContactGroupListResponse.ValidateAll() if the designated
// constraints aren't met.
type ContactGroupListResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupListResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupListResponseMultiError) AllErrors() []error { return m }

// ContactGroupListResponseValidationError is the validation error returned by
// ContactGroupListResponse.Validate if the designated constraints aren't met.
type ContactGroupListResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupListResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupListResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupListResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupListResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupListResponseValidationError) ErrorName() string {
	return "ContactGroupListResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupListResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupListResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupListResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupListResponseValidationError{}

// Validate checks the field values on ContactGroupSaveRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupSaveRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupSaveRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupSaveRequestMultiError, or nil if none found.
func (m *ContactGroupSaveRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupSaveRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetItems() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ContactGroupSaveRequestValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ContactGroupSaveRequestValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactGroupSaveRequestValidationError{
					field:  fmt.Sprintf("Items[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ContactGroupSaveRequestMultiError(errors)
	}

	return nil
}

// ContactGroupSaveRequestMultiError is an error wrapping multiple validation
// errors returned by ContactGroupSaveRequest.ValidateAll() if the designated
// constraints aren't met.
type ContactGroupSaveRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupSaveRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupSaveRequestMultiError) AllErrors() []error { return m }

// ContactGroupSaveRequestValidationError is the validation error returned by
// ContactGroupSaveRequest.Validate if the designated constraints aren't met.
type ContactGroupSaveRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupSaveRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupSaveRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupSaveRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupSaveRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupSaveRequestValidationError) ErrorName() string {
	return "ContactGroupSaveRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupSaveRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupSaveRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupSaveRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupSaveRequestValidationError{}

// Validate checks the field values on ContactGroupSaveResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupSaveResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupSaveResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupSaveResponseMultiError, or nil if none found.
func (m *ContactGroupSaveResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupSaveResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ContactGroupSaveResponseMultiError(errors)
	}

	return nil
}

// ContactGroupSaveResponseMultiError is an error wrapping multiple validation
// errors returned by ContactGroupSaveResponse.ValidateAll() if the designated
// constraints aren't met.
type ContactGroupSaveResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupSaveResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupSaveResponseMultiError) AllErrors() []error { return m }

// ContactGroupSaveResponseValidationError is the validation error returned by
// ContactGroupSaveResponse.Validate if the designated constraints aren't met.
type ContactGroupSaveResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupSaveResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupSaveResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupSaveResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupSaveResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupSaveResponseValidationError) ErrorName() string {
	return "ContactGroupSaveResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupSaveResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupSaveResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupSaveResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupSaveResponseValidationError{}

// Validate checks the field values on ContactListResponse_Item with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactListResponse_Item) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactListResponse_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactListResponse_ItemMultiError, or nil if none found.
func (m *ContactListResponse_Item) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactListResponse_Item) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Username

	// no validation rules for Avatar

	// no validation rules for Name

	// no validation rules for Surname

	// no validation rules for Gender

	// no validation rules for About

	// no validation rules for IsOnline

	// no validation rules for GroupId

	if len(errors) > 0 {
		return ContactListResponse_ItemMultiError(errors)
	}

	return nil
}

// ContactListResponse_ItemMultiError is an error wrapping multiple validation
// errors returned by ContactListResponse_Item.ValidateAll() if the designated
// constraints aren't met.
type ContactListResponse_ItemMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactListResponse_ItemMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactListResponse_ItemMultiError) AllErrors() []error { return m }

// ContactListResponse_ItemValidationError is the validation error returned by
// ContactListResponse_Item.Validate if the designated constraints aren't met.
type ContactListResponse_ItemValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactListResponse_ItemValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactListResponse_ItemValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactListResponse_ItemValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactListResponse_ItemValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactListResponse_ItemValidationError) ErrorName() string {
	return "ContactListResponse_ItemValidationError"
}

// Error satisfies the builtin error interface
func (e ContactListResponse_ItemValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactListResponse_Item.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactListResponse_ItemValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactListResponse_ItemValidationError{}

// Validate checks the field values on ContactRequestListResponse_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactRequestListResponse_Item) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequestListResponse_Item with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// ContactRequestListResponse_ItemMultiError, or nil if none found.
func (m *ContactRequestListResponse_Item) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequestListResponse_Item) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for UserId

	// no validation rules for FriendId

	// no validation rules for Username

	// no validation rules for Avatar

	// no validation rules for Name

	// no validation rules for Surname

	if len(errors) > 0 {
		return ContactRequestListResponse_ItemMultiError(errors)
	}

	return nil
}

// ContactRequestListResponse_ItemMultiError is an error wrapping multiple
// validation errors returned by ContactRequestListResponse_Item.ValidateAll()
// if the designated constraints aren't met.
type ContactRequestListResponse_ItemMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestListResponse_ItemMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestListResponse_ItemMultiError) AllErrors() []error { return m }

// ContactRequestListResponse_ItemValidationError is the validation error
// returned by ContactRequestListResponse_Item.Validate if the designated
// constraints aren't met.
type ContactRequestListResponse_ItemValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestListResponse_ItemValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestListResponse_ItemValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestListResponse_ItemValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestListResponse_ItemValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestListResponse_ItemValidationError) ErrorName() string {
	return "ContactRequestListResponse_ItemValidationError"
}

// Error satisfies the builtin error interface
func (e ContactRequestListResponse_ItemValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequestListResponse_Item.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestListResponse_ItemValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestListResponse_ItemValidationError{}

// Validate checks the field values on ContactGroupSortRequest_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupSortRequest_Item) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupSortRequest_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupSortRequest_ItemMultiError, or nil if none found.
func (m *ContactGroupSortRequest_Item) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupSortRequest_Item) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Sort

	if len(errors) > 0 {
		return ContactGroupSortRequest_ItemMultiError(errors)
	}

	return nil
}

// ContactGroupSortRequest_ItemMultiError is an error wrapping multiple
// validation errors returned by ContactGroupSortRequest_Item.ValidateAll() if
// the designated constraints aren't met.
type ContactGroupSortRequest_ItemMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupSortRequest_ItemMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupSortRequest_ItemMultiError) AllErrors() []error { return m }

// ContactGroupSortRequest_ItemValidationError is the validation error returned
// by ContactGroupSortRequest_Item.Validate if the designated constraints
// aren't met.
type ContactGroupSortRequest_ItemValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupSortRequest_ItemValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupSortRequest_ItemValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupSortRequest_ItemValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupSortRequest_ItemValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupSortRequest_ItemValidationError) ErrorName() string {
	return "ContactGroupSortRequest_ItemValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupSortRequest_ItemValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupSortRequest_Item.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupSortRequest_ItemValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupSortRequest_ItemValidationError{}

// Validate checks the field values on ContactGroupListResponse_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupListResponse_Item) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupListResponse_Item with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// ContactGroupListResponse_ItemMultiError, or nil if none found.
func (m *ContactGroupListResponse_Item) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupListResponse_Item) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for Count

	// no validation rules for Sort

	if len(errors) > 0 {
		return ContactGroupListResponse_ItemMultiError(errors)
	}

	return nil
}

// ContactGroupListResponse_ItemMultiError is an error wrapping multiple
// validation errors returned by ContactGroupListResponse_Item.ValidateAll()
// if the designated constraints aren't met.
type ContactGroupListResponse_ItemMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupListResponse_ItemMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupListResponse_ItemMultiError) AllErrors() []error { return m }

// ContactGroupListResponse_ItemValidationError is the validation error
// returned by ContactGroupListResponse_Item.Validate if the designated
// constraints aren't met.
type ContactGroupListResponse_ItemValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupListResponse_ItemValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupListResponse_ItemValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupListResponse_ItemValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupListResponse_ItemValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupListResponse_ItemValidationError) ErrorName() string {
	return "ContactGroupListResponse_ItemValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupListResponse_ItemValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupListResponse_Item.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupListResponse_ItemValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupListResponse_ItemValidationError{}

// Validate checks the field values on ContactGroupSaveRequest_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ContactGroupSaveRequest_Item) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactGroupSaveRequest_Item with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactGroupSaveRequest_ItemMultiError, or nil if none found.
func (m *ContactGroupSaveRequest_Item) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactGroupSaveRequest_Item) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Sort

	// no validation rules for Name

	if len(errors) > 0 {
		return ContactGroupSaveRequest_ItemMultiError(errors)
	}

	return nil
}

// ContactGroupSaveRequest_ItemMultiError is an error wrapping multiple
// validation errors returned by ContactGroupSaveRequest_Item.ValidateAll() if
// the designated constraints aren't met.
type ContactGroupSaveRequest_ItemMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactGroupSaveRequest_ItemMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactGroupSaveRequest_ItemMultiError) AllErrors() []error { return m }

// ContactGroupSaveRequest_ItemValidationError is the validation error returned
// by ContactGroupSaveRequest_Item.Validate if the designated constraints
// aren't met.
type ContactGroupSaveRequest_ItemValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactGroupSaveRequest_ItemValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactGroupSaveRequest_ItemValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactGroupSaveRequest_ItemValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactGroupSaveRequest_ItemValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactGroupSaveRequest_ItemValidationError) ErrorName() string {
	return "ContactGroupSaveRequest_ItemValidationError"
}

// Error satisfies the builtin error interface
func (e ContactGroupSaveRequest_ItemValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactGroupSaveRequest_Item.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactGroupSaveRequest_ItemValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactGroupSaveRequest_ItemValidationError{}
